hudi-utilities/src/main/java/org/apache/hudi/utilities/keygen/TimestampBasedKeyGenerator.java

```diff
@@ -102,7 +109,12 @@ public HoodieKey getKey(GenericRecord record) {
             "Unexpected type for partition field: " + partitionVal.getClass().getName());
       }
 
-      return new HoodieKey(DataSourceUtils.getNestedFieldValAsString(record, recordKeyField),
+      return new HoodieKey(
+          fields.stream()
+              .map(
+                  recordKeyField ->
+                      DataSourceUtils.getNestedFieldValAsString(record, recordKeyField))
+              .collect(Collectors.joining(".")),
```

Comment:
We are joining by "." here. Is there any assumption here ? Wanted to see if this is generic enough to be put it in this class.

Reply:
I think it can be anything. 
HoodieKey(String recordKey, String partitionPath)

so I'm just creating a long key: bla.bla1.bla2. Can be bla:bla1:bla2

Reply:
Hi I think this kind of key creates ambiguity  refer this PR for details https://github.com/apache/incubator-hudi/pull/728

Reply:
@bvaradar any thoughts? lets make a call on this PR?

Reply:
@jaimin-shah : Wondering if there is a necessity to decode the generated record key. If this is a one-way concatenation of fields to record-key, it should be fine. right ? As we are storing individual fields that constitute the record key separately, there wont be any need to decode this record key and  ambiguity during decoding should be ok. Isn't it ?

Reply:
@bvaradar By ambiguity I meant two different records having same key. For example  US, .A => US..A and US. , A => US..A Keeping recordKeyField as part of key resolves this. Although I agree these kind of cases are quite rare.

Reply:
Hey, not sure I understood the edge case example. Here is how we are using it:
We have an object with id and a version. Version suppose to monotonically increase, and we want to be able to dedub based on key and the version (there should be at most 1 object with the same id and version).

Our config looks like:
hoodie.datasource.write.recordkey.field=object.id,object.version for insert table and

What can go wrong with this one?

Reply:
Hi @afilipchik for your use case I don't think there will be any problem but there can be problem when there no restrictions on recordKeyField ( e.g. Version suppose to monotonically increase ). You can take a look at  https://github.com/apache/incubator-hudi/blob/master/hudi-spark/src/main/java/org/apache/hudi/ComplexKeyGenerator.java it has generic implementation of complex key similar requirement.

Reply:
this link is not available.